mayastor/src/bdev/nexus/nexus_io_subsystem.rs

```diff
@@ -0,0 +1,182 @@
+                // Wait till the subsystem has completed transition and retry
+                // operation.
+                Err(NexusPauseState::Unpausing)
+                | Err(NexusPauseState::Pausing) => {
+                    trace!(nexus=%self.name, "nexus is in intermediate state, deferring Pause operation");
+                    let (s, r) = oneshot::channel::<i32>();
+                    self.pause_waiters.push_back(s);
+                    r.await.unwrap();
+                    trace!(nexus=%self.name, "nexus completed state transition, retrying Pause operation");
+                }
+                _ => {
+                    panic!("Corrupted I/O subsystem state");
+                }
+            };
+        }
+
+        // Resume one waiter in case there are any.
+        if !self.pause_waiters.is_empty() {
+            let w = self.pause_waiters.pop_front().unwrap();
+            trace!(nexus=%self.name, "resuming the first Pause waiter");
+            w.send(0).expect("I/O subsystem pause waiter disappeared");
+        }
+
+        trace!(?self.name, "nexus I/O paused");
+        Ok(())
+    }
+
+    /// Resume IO to the bdev.
+    /// Note: in order to handle concurrent resumes properly, this function must
+    /// be called only from the master core.
+    pub async fn resume(&mut self) -> Result<(), NexusError> {
+        assert_eq!(Cores::current(), Cores::first());
+
+        trace!(?self.name, "resuming nexus I/O");
+
+        loop {
+            match self.pause_state.load() {
+                // Already unpaused, bail out.
+                NexusPauseState::Unpaused => {
+                    break;
+                }
+                // Simultaneous pausing/unpausing: wait till the subsystem has
+                // completed transition and retry operation.
+                NexusPauseState::Pausing | NexusPauseState::Unpausing => {
+                    trace!(?self.name, "nexus is in intermediate state, deferring Resume operation");
+                    let (s, r) = oneshot::channel::<i32>();
+                    self.pause_waiters.push_back(s);
+                    r.await.unwrap();
+                    trace!(?self.name, "completed state transition, retrying Resume operation");
+                }
+                // Unpause the subsystem, taking into account the overall number
+                // of pauses.
+                NexusPauseState::Paused => {
+                    let v = self.pause_cnt.fetch_sub(1, Ordering::SeqCst);
+                    // In case the last pause discarded, resume the subsystem.
+                    if v == 1 {
+                        if let Some(subsystem) =
+                            NvmfSubsystem::nqn_lookup(&self.name)
+                        {
+                            self.pause_state.store(NexusPauseState::Unpausing);
+                            trace!(nexus=%self.name, nqn=%subsystem.get_nqn(), "resuming subsystem");
+                            subsystem.resume().await.unwrap();
```

Comment:
when can this return an error and do we need to cater for that?

Reply:
We just call libspdk here. In theory, when there is a concurrent 'suspend()' call to spdk, this can panic, as it's not possible for SPKD to handle these 2 operations at the same time.
But this fix is about preventing such a situation in the first place by serializing calls to spdk. :-)